### ISSUE TITLE
Dynamic proto3 maps do not unmarshal from text format

### ISSUE THREAD

**Issue by jkinkead** [jhump/protoreflect#71]
A map-typed field does not correctly unmarshal from text; it seems to ignore all input.

Given a message like:
```proto
message TestMessage {
  map<string, int32> mapValue = 1;
}
```
A dynamic message should be able to parse the text:
```
mapValue: <key: "bar" value: 1>
mapValue: <key: "foo" value: 2>
```
Passing that text to `proto.UnmarshalText` with a dynamic message produces no change in the saved map data (but also returns no error).

Note that serialization (`proto.MarshalText`) produces correct output when given a message with a map value set.

**Comment by jhump** [comment]
There are test cases for exactly this that pass. But, like the issue you discovered in `protoparse`, this only impacts proto3 messages.

It turns out that even the `PutMapField` method of `dynamic.Message` silently fails when trying to insert the first value into a map field :/  (same root cause as this).

The underlying issue is that there is some logic that tries to initialize the field value to an empty map, and then put the new entry into that map. That doesn't work for proto3 because we don't bother to store zero values (in proto3, zero values, such as an empty map for a map field, are indistinguishable from absent). I've found and fixed a similar bug in de-serialization, but missed this code path.

Fix is in #72.

Sorry you're finding issues with proto3. Most of the test files use proto2 since things like default values and extensions are only supported in proto2. I apparently need to address that gap by adding more proto3 test cases.

**Comment by jkinkead** [comment]
No worries, thanks for the quick fixes!
